### PR TITLE
[Enhancement] Support Query cache in shared_data

### PR DESCRIPTION
**Pull request by smartlxh**
## Why I'm doing:
Support Query cache in shared_data
## What I'm doing:
1. Support `lake::TabletReader`  read  delta_rowsets
Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Review by yingtingdong — DISMISSED** [review]

File: be/src/storage/lake/tablet_manager.cpp [hunk 1]
```diff
@@ -623,4 +623,35 @@ StatusOr<SegmentPtr> TabletManager::load_segment(const FileInfo& segment_info, i
...
+StatusOr<TabletAndRowsets> TabletManager::capture_tablet_and_rowsets(int64_t tablet_id, int64_t from_version,
+                                                                     int64_t to_version) {
+    auto tablet_ptr = std::make_shared<Tablet>(this, tablet_id);
+
+    std::vector<std::shared_ptr<BaseRowset>> rowsets;
+
+    std::vector<RowsetPtr> from_rowsets;
+    std::vector<RowsetPtr> to_rowsets;
+    if (from_version != 0) {
+        auto tablet_meta_from_version = get_tablet_metadata(tablet_id, from_version).value();
```
Comment (wyb) [inline be/src/storage/lake/tablet_manager.cpp:635]: `from_version` metadata may be not exist.
or other errors.

File: be/src/storage/lake/tablet_manager.cpp [hunk 2]
```diff
@@ -623,4 +623,35 @@ StatusOr<SegmentPtr> TabletManager::load_segment(const FileInfo& segment_info, i
...
+    auto tablet_meta_to_version = get_tablet_metadata(tablet_id, to_version).value();
+    to_rowsets = Rowset::get_rowsets(this, tablet_meta_to_version);
+
+    std::unordered_map<int64_t, std::shared_ptr<Rowset>> distinct_rowset;
+    for (const auto& rowset : from_rowsets) {
+        distinct_rowset[rowset->id()] = std::move(rowset);
+    }
+
+    for (const auto& rowset : to_rowsets) {
+        if (distinct_rowset.find(rowset->id()) != distinct_rowset.end()) {
```
Comment (wyb) [inline be/src/storage/lake/tablet_manager.cpp:648]: how about tablet metadata generated by compaction or after the compaction?
should get the real delta rowsets.

**Review by satanson — DISMISSED** [review]

File: be/src/exec/query_cache/cache_operator.cpp [hunk 3]
```diff
@@ -295,17 +329,50 @@ void CacheOperator::_handle_stale_cache_value_for_pk(int64_t tablet_id, starrock
...
+        base_rowsets = std::move(rowsets);
     }
-    auto& [tablet, rowsets, rowsets_acq_rel] = status.value();
+
     const auto snapshot_version = cache_value.version;
     bool can_pickup_delta_rowsets = false;
     bool exists_non_empty_delta_rowsets = false;
-    for (auto& rs : rowsets) {
+
+    if (_cache_param.is_lake && !base_rowsets.empty()) {
```
Comment (wyb) [inline be/src/exec/query_cache/cache_operator.cpp:370]: if `base_rowsets` is empty, is `PLBS_HIT_TOTAL` definitely right?

File: fe/fe-core/src/main/java/com/starrocks/planner/FragmentNormalizer.java [hunk 4]
```diff
@@ -330,6 +331,10 @@ public boolean computeDigest(PlanNode cachePointNode) {
             cacheParam.setCan_use_multiversion(canUseMultiVersion);
             cacheParam.setKeys_type(keysType.toThrift());
             cacheParam.setCached_plan_node_ids(cachedPlanNodeIds);
+            if (RunMode.getCurrentRunMode() == RunMode.SHARED_DATA) {
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/planner/FragmentNormalizer.java:334]:
```suggestion
            if (RunMode.isSharedDataMode()) {
```

**Review by satanson — DISMISSED** [review]

File: be/src/exec/query_cache/cache_operator.cpp [hunk 5]
```diff
@@ -295,17 +329,45 @@ void CacheOperator::_handle_stale_cache_value_for_pk(int64_t tablet_id, starrock
...
     // At the present, PRIMARY_KEYS can not support merge-on-read, so we can not merge stale cache values and delta
     // rowsets. Capturing delta rowsets is meaningless and unsupported, thus we capture all rowsets of the PK tablet.
-    auto status = StorageEngine::instance()->tablet_manager()->capture_tablet_and_rowsets(tablet_id, 0, version);
-    if (!status.ok()) {
-        buffer->state = PLBS_MISS;
-        buffer->cached_version = 0;
-        return;
+    std::shared_ptr<BaseTablet> base_tablet;
+    std::vector<BaseRowsetSharedPtr> base_rowsets;
+    RowsetsAcqRelPtr rowsets_acq_rel = nullptr;
```
Comment (wyb) [inline be/src/exec/query_cache/cache_operator.cpp:334]: `rowsets_acq_rel` is not used

**Review by wyb — DISMISSED** [review]

**Review by satanson — DISMISSED** [review]

**Review by wyb — DISMISSED** [review]

**Review by satanson — DISMISSED** [review]

**Review by wyb — APPROVED** [review]

**Review by satanson — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Pull request merged**